File: lab4-primit/haskell/errorsEnum.hs
data Error = FP_Rounding | FP_Overflow | FP_Underflow | Int_Overflow
    deriving (Show, Read, Enum, Bounded)

data Result = ABitDifferent | Infinity | Zero | VeryDifferent
    deriving (Show, Read, Enum, Bounded)

error2Result :: Error -> Result
error2Result FP_Rounding  = ABitDifferent
error2Result FP_Overflow  = Infinity
error2Result FP_Underflow = Zero
error2Result Int_Overflow = VeryDifferent

result2Error :: Result -> Error
result2Error ABitDifferent = FP_Rounding
result2Error Infinity      = FP_Overflow
result2Error Zero         = FP_Underflow
result2Error VeryDifferent = Int_Overflow

allErrors :: [Error]
allErrors = [minBound .. maxBound]

allResults :: [Result]
allResults = [minBound .. maxBound]

main :: IO ()
main = do
    putStrLn ("known errors = " ++ show allErrors)
    error <- getLine
    putStrLn (show error ++ " results in: " ++ show (error2Result (read error)))

    putStrLn ("known results = " ++ show allResults)
    result <- getLine
    putStrLn (show result ++ " results from: " ++ show (result2Error (read result)))

File: lab4-primit/javascript/ErrorsEnum.js
const Error_enumobj = {
    FP_ROUNDING: "FP_ROUNDING",
    FP_OVERFLOW: "FP_OVERFLOW",
    FP_UNDERFLOW: "FP_UNDERFLOW",
    INT_OVERFLOW: "INT_OVERFLOW"
};

const Result_enumobj = {
    A_BIT_DIFFERENT: "A_BIT_DIFFERENT",
    INFINITY: "INFINITY",
    ZERO: "ZERO",
    VERY_DIFFERENT: "VERY_DIFFERENT"
};

function error2Result(err) {
    switch (err) {
        case Error_enumobj.FP_ROUNDING:
            return Result_enumobj.A_BIT_DIFFERENT;
        case Error_enumobj.FP_OVERFLOW:
            return Result_enumobj.INFINITY;
        case Error_enumobj.FP_UNDERFLOW:
            return Result_enumobj.ZERO;
        case Error_enumobj.INT_OVERFLOW:
            return Result_enumobj.VERY_DIFFERENT;
        default:
            return null;
    }
}

function result2Error(result) {
    switch (result) {
        case Result_enumobj.A_BIT_DIFFERENT:
            return Error_enumobj.FP_ROUNDING;
        case Result_enumobj.INFINITY:
            return Error_enumobj.FP_OVERFLOW;
        case Result_enumobj.ZERO:
            return Error_enumobj.FP_UNDERFLOW;
        case Result_enumobj.VERY_DIFFERENT:
            return Error_enumobj.INT_OVERFLOW;
        default:
            return null;
    }
}

// Display known errors and results
console.log("Error list:", Object.values(Error_enumobj));
console.log("Result list:", Object.values(Result_enumobj));

// Example usage
let input = "FP_OVERFLOW"; // Change this to test different cases
if (Object.values(Error_enumobj).includes(input)) {
    let result = error2Result(input);
    console.log(input + " results in " + result);
} else if (Object.values(Result_enumobj).includes(input)) {
    let err = result2Error(input);
    console.log(input + " results from " + err);
} else {
    console.log("Invalid input");
}
